### PR TITLE
Add docker build CICD

### PR DESCRIPTION
**Pull request by mrchtr**
Add script to build and push the custom component to `ghcr.io`
Set the label to `org.opencontainers.image.source=https://github.com/ml6team/fondant-usecase-RAG`

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+name: Build dev images
+
+on:
+  push:
+    branches:
+      - '*'
```
Comment (mrchtr) [inline .github/workflows/build.yaml:6]: Should be set to main before we merge it.

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        uses: docker/setup-buildx-action@v2
+
+      - name: Set buildx alias
+        run: docker buildx install
+
+      - name: Install docker pushrm
+        run: |
+          sudo wget https://github.com/christian-korneck/docker-pushrm/releases/download/v1.9.0/docker-pushrm_linux_amd64 -O /usr/libexec/docker/cli-plugins/docker-pushrm
+          sudo chmod +x /usr/libexec/docker/cli-plugins/docker-pushrm
+          docker pushrm --help
```
Comment (RobbeSneyders) [inline .github/workflows/build.yaml:26]: Does this work with github container registry?

I would be fine with leaving this out for the example images.

File: scripts/build_components.sh [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    cache_name=ghcr.io/${namespace}/${BASENAME}:build-cache
+    echo "Caching from/to ${cache_name}"
+    args+=(--cache-to "type=registry,ref=${cache_name}")
+    args+=(--cache-from "type=registry,ref=${cache_name}")
+  fi
+
+  echo "Freezing Fondant dependency version to ${tags[0]}"
+
+  docker build --push "${args[@]}" \
+   --build-arg="FONDANT_VERSION=${tags[0]}" \
```
Comment (RobbeSneyders) [inline scripts/build_components.sh:103]: Custom docker files don't contain the `FONDANT_VERSION` build argument.

File: scripts/build_components.sh [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  done
+
+  # Add cache arguments if caching is enabled
+  if [ "$caching" = true ] ; then
+
+    cache_name=ghcr.io/${namespace}/${BASENAME}:build-cache
+    echo "Caching from/to ${cache_name}"
+    args+=(--cache-to "type=registry,ref=${cache_name}")
+    args+=(--cache-from "type=registry,ref=${cache_name}")
+  fi
```
Comment (RobbeSneyders) [inline scripts/build_components.sh:98]: I would leave out caching for custom images. It adds a lot of complexity while it's mostly useful when you have a lot of images.

File: scripts/build_components.sh [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  echo "Tagging image with following tags:"
+  for tag in "${tags[@]}"; do
+    full_image_name=ghcr.io/${namespace}/${BASENAME}:${tag}
+    echo "$full_image_name"
+    full_image_names+=("$full_image_name")
+  done
+
+  echo "Updating the image version in the fondant_component.yaml with:"
+  echo "${full_image_names[0]}"
+  sed -i -e "s|^image: .*|image: ${full_image_names[0]}|" fondant_component.yaml
```
Comment (RobbeSneyders) [inline scripts/build_components.sh:82]: This is not needed for CI/CD, and for other usage, their's now the `fondant build` command.

File: scripts/build_components.sh [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+# Loop through all subdirectories
+for dir in "${components_to_build[@]}"; do
+  pushd "$dir"
+  BASENAME=${dir%/}
+  BASENAME=${BASENAME##*/}
+
+  full_image_names=()
+  echo "Tagging image with following tags:"
+  for tag in "${tags[@]}"; do
+    full_image_name=ghcr.io/${namespace}/${BASENAME}:${tag}
```
Comment (RobbeSneyders) [inline scripts/build_components.sh:75]: We might want to keep the registry parametrized so users can easily switch to a different registry.

File: scripts/build_components.sh [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+function usage {
+  echo "Usage: $0 [options]"
+  echo "Options:"
+  echo "  -t,  --tag <value>                 Tag to add to image, repeatable
+                                             The first tag is set in the component specifications"
+  echo "  -c,  --cache <value>               Use registry caching when building the components (default:false)"
+  echo "  -d,  --components-dir <value>      Directory containing components to build as subdirectories.
+                                             The path should be relative to the root directory (default:components)"
+  echo "  -r,  --registry <value>            The docker registry prefix to use (default: null for DockerHub)"
+  echo "  -n,  --namespace <value>           The DockerHub namespace for the built images (default: fndnt)"
```
Comment (RobbeSneyders) [inline scripts/build_components.sh:13]: Some outdated documentation here

File: .github/workflows/build.yaml (hunk 2 above)
Comment (RobbeSneyders) [reply .github/workflows/build.yaml:26]: Seems like it's not used in the script.

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @mrchtr,

I think there's a few things in the script we need / might want to change for a script in the example repositories, see my comments. Maybe we should leverage `fondant build` instead?

File: scripts/build_components.sh [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+function usage {
+  echo "Usage: $0 [options]"
+  echo "Options:"
+  echo "  -t,  --tag <value>                 Tag to add to image
+                                             The tag is set in the component specifications"
+  echo "  -d,  --components-dir <value>      Directory containing components to build as subdirectories.
+                                             The path should be relative to the root directory (default:src/components)"
+  echo "  -r,  --registry <value>            The docker registry prefix to use (default: null for DockerHub)"
+  echo "  -n,  --namespace <value>           The DockerHub namespace for the built images (default: ml6team)"
```
Comment (RobbeSneyders) [inline scripts/build_components.sh:12]: You still mention DockerHub here twice.

**Review by RobbeSneyders — APPROVED** [review]
Thanks @mrchtr!

Seems to be working :+1: 
Can you update the branch name in the workflow before merging?

**Pull request merged**